src/DIRAC/DataManagementSystem/Agent/RequestOperations/test/Test_ReplicateAndRegister.py

```diff
@@ -0,0 +1,690 @@
+    """If we have a disk and tape replicas, but specify the tape replicas as source,
+    we want the tape one
+    """
+
+    lfn = os.path.join("/", MARKER_WITH_REPLICAS, "Disk1", "Tape1", END_MARKER)
+
+    def initialize_test(self):
+        rmsFile = super().initialize_test()
+        self.filterReplicasKwargs = {"opSources": ["Tape1"]}
+        return rmsFile
+
+    def check_result(self, res):
+        stateAfter = self.rmsFile.toJSON()
+
+        assert res["OK"]
+        assert res["Value"]["Valid"] == ["Tape1"]
+
+        _compareFileAttr(self.stateBefore, stateAfter, {"ChecksumType", "Checksum"})
+
+
+@runAsPytest
+class test_case_noActiveReplicas(BaseTestCase):
+    """The file has no replicas available"""
+
+    lfn = NO_AVAILABLE_REPLICAS
+
+    def check_result(self, res):
+        stateAfter = self.rmsFile.toJSON()
+
+        assert res["OK"]
+        # When no replicas are available, we just get "None"
+        assert res["Value"]["NoReplicas"] == [None]
+
+        # We should have added the catalog checksum
+        _compareFileAttr(self.stateBefore, stateAfter, {"ChecksumType", "Checksum"})
+
+        assert self.rmsFile.Checksum == adler32(self.rmsFile.LFN.encode())
+
+
+@runAsPytest
+class test_case_errorGettingSEMetadata(BaseTestCase):
+    """When we have a complete failure of getting the file metadata from SE,
+    the files goes to "NoMetadata"
+    """
+
+    lfn = ERROR_GETTING_SE_METADATA
+
+    def check_result(self, res):
+        stateAfter = self.rmsFile.toJSON()
+
+        assert res["OK"]
+
+        # We should have a Valid replicas
+        assert DEFAULT_SE in res["Value"]["NoMetadata"]
+
+        # We got the checksum from the FC
+
+        _compareFileAttr(self.stateBefore, stateAfter, {"ChecksumType", "Checksum"})
+
+
+@runAsPytest
+class test_case_fileRegisteredButDoesNotPhysicalyExist(BaseTestCase):
+    """File has two replicas registered, but one of them does not exist"""
+
+    # Two replicas
+    lfnParts = ["/", MARKER_WITH_REPLICAS, "Disk1", "Disk2", END_MARKER]
+    # but it doesn't exist
+    lfnParts += [MARKER_BAD_SE_NOT_EXIST, "Disk2", END_MARKER]
+
+    lfn = os.path.join(*lfnParts)
+
+    def check_result(self, res):
+        stateAfter = self.rmsFile.toJSON()
+
+        assert res["OK"]
+
+        assert res["Value"]["Valid"] == ["Disk1"]
+        assert res["Value"]["NoReplicas"] == ["Disk2"]
+        # We should have added the catalog checksum
+        _compareFileAttr(self.stateBefore, stateAfter, {"ChecksumType", "Checksum"})
+
+
+@pytest.mark.xfail(reason="filterReplicas does not support bulk")
+@runAsPytest
+class test_all(BaseTestCase):
+    """
+    This test re-run all the other tests, but by giving all the LFNs at once.
+    I wanted to make filterReplicas a bulk method (not a single rmsFile, but multiple),
+    so this would have been a neat way of testing it.
+    I've decided not to do it in the end, but I keep the test here, again for doc purpose
```

Comment:
Do you want to tell us why? Better to document if there were reasons to not do it, instead of discovering the same reason in a few years.

Reply:
oh just that it would change a lot the logic of the method and completely change the interface. I actually intended this PR as a small one :D